File: src/Display.cpp
#include "Display.h"

std::map<int, Cairo::Format> bpp_to_format{
    {32, Cairo::Format::FORMAT_ARGB32},
    {16, Cairo::Format::FORMAT_RGB16_565}};

int get_compass_line_length(int deg)
{
    if (deg % 45 == 0)
        return 20;
    if (deg % 10 == 0)
        return 15;
    if (deg % 5 == 0)
        return 10;
    return 0;
}

template <typename T>
std::string to_string_with_precision(const T a_value, const int n = 6)
{
    std::ostringstream out;
    out.precision(n);
    out << std::fixed << a_value;
    return out.str();
}

const std::string currentDateTime()
{
    setenv("TZ", "PST8PDT", 1);
    tzset();

    time_t now = time(0);
    struct tm tstruct;
    char buf[80];
    tstruct = *localtime(&now);
    // Visit http://en.cppreference.com/w/cpp/chrono/c/strftime
    // for more information about date/time format
    strftime(buf, sizeof(buf), "%a, %b %e, %l:%M:%S %p", &tstruct);

    return buf;
}

Display::Display()
{
    surface = Cairo::ImageSurface::create((unsigned char *)fb.buff, bpp_to_format[fb.fb_info.var.bits_per_pixel],
                                          fb.fb_info.var.xres, fb.fb_info.var.yres, fb.fb_info.fix.line_length);

    ctx = Cairo::Context::create(surface);

    width = surface->get_width();
    height = surface->get_height();

    str_out << std::setprecision(2);
};

void Display::put_text(double x, double y)
{
    ctx->move_to(x, y);
    ctx->show_text(str_out.str());
    str_out.str("");
    str_out.clear();
};

void Display::update(GPS &gps, Data &data)
{
    ctx->set_identity_matrix();

    ctx->rectangle(0, 0, surface->get_width(), surface->get_height());
    ctx->set_source_rgb(0, 0, 0);
    ctx->fill();

    ctx->select_font_face("serif",
                          Cairo::FontSlant::FONT_SLANT_NORMAL,
                          Cairo::FontWeight::FONT_WEIGHT_BOLD);

    double font_size = 40;
    ctx->set_font_size(font_size);

    ctx->set_source_rgb(1, 1, 1);

    // start_screen(gps, data);
    data_screen(gps, data);

    // top_bar(gps);

    // str_out << "Latitude: " << gps.lat;
    // put_text(20, line_height);

    // str_out << "Longitude: " << gps.lon;
    // put_text(20, line_height * 2);

    // str_out << "Speed: " << gps.speed << " kts";
    // put_text(20, line_height * 3);

    // str_out << "CMG: " << gps.heading;
    // put_text(20, line_height * 4);

    // ctx->set_identity_matrix();
    // a += 0.005;
    // draw_compass(a * 10);

    fb.flip();
};

void Display::draw_compass(double value)
{
    ctx->set_identity_matrix();

    ctx->set_source_rgb(1, 1, 1);

    int compass_arc = 360 / 8;
    int compass_rad = 150;

    ctx->begin_new_path();
    ctx->translate(width / 2, height / 2); // position
    ctx->arc_negative(0, 0, compass_rad, DEG_RAD * (compass_arc - 90), DEG_RAD * (-compass_arc - 90));
    ctx->stroke();

    ctx->move_to(0, 0);

    ctx->set_font_size(30);

    std::string label = to_string_with_precision(value, 1);
    Cairo::TextExtents extents;
    ctx->get_text_extents(label, extents);
    ctx->rel_move_to(-extents.width / 2, -compass_rad + extents.height / 2 + 35);
    ctx->text_path(label);
    ctx->fill_preserve();

    ctx->set_font_size(20);

    int offset = (int)value % 5;
    ctx->rotate(DEG_RAD * (-offset - 45));
    for (int true_deg = -compass_arc - offset; true_deg <= compass_arc - offset; true_deg += 5)
    {
        int deg = true_deg + (int)value + 5;

        ctx->move_to(0, 0);
        ctx->rotate(DEG_RAD * 5);
        ctx->rel_move_to(0, -compass_rad - 30);

        int line_length = get_compass_line_length(deg);
        ctx->rel_line_to(0, line_length);
        ctx->stroke_preserve();

        if (deg % 20 == 0)
        {
            ctx->rel_move_to(0, -15);

            std::string label = std::to_string(deg % 360);

            ctx->get_text_extents(label, extents);
            ctx->rel_move_to(-extents.width / 2, -extents.height / 2);
            ctx->text_path(label);
        }

        ctx->fill();
    }
}

void Display::label_bar(double divider_y, double divider_height, std::vector<std::string> labels)
{
    double cell_width = (double)width / labels.size();
    double text_width = cell_width * 0.75;

    ctx->move_to(0, divider_y);
    ctx->line_to(width, divider_y);

    ctx->move_to(0, divider_y + divider_height);
    ctx->line_to(width, divider_y + divider_height);

    Cairo::TextExtents extents;

    double font_size = 100;

    for (unsigned int i = 1; i < labels.size(); i++)
    {
        ctx->move_to(i * cell_width, divider_y + divider_height);
        ctx->line_to(i * cell_width, divider_y);

        auto label = labels.at(i);

        ctx->set_font_size(font_size);
        ctx->get_text_extents(label, extents);
        double new_font_size = font_size / extents.width * text_width;
        if (new_font_size < font_size)
            font_size = new_font_size;
    }
    ctx->stroke();

    ctx->set_font_size(font_size);

    for (unsigned int i = 0; i < labels.size(); i++)
    {
        auto label = labels.at(i);

        ctx->get_text_extents(label, extents);
        ctx->move_to(cell_width * i + (cell_width - extents.width) * 0.5, divider_y + (divider_height + extents.height) * 0.5);
        ctx->text_path(label);
    }

    ctx->fill();
}
void Display::data_screen(GPS &gps, Data &data)
{
    label_bar(0, height * 0.1, {
                                   "Sats: " + std::to_string(gps.satellites_used) + "/" + std::to_string(gps.satellites_visible),
                                   "Accuracy: " + to_string_with_precision(gps.accuracy, 2) + "m",
                                   currentDateTime(),
                               });
    draw_compass(data.speed);
}
void Display::start_screen(GPS &gps, Data &data)
{

    // TOP INFO

    label_bar(0, height * 0.1, {
                                   "Sats: " + std::to_string(gps.satellites_used) + "/" + std::to_string(gps.satellites_visible),
                                   "Accuracy: " + to_string_with_precision(gps.accuracy, 2) + "m",
                                   currentDateTime(),
                               });

    // BOTTOM INFO

    std::string time_label = "5:00";

    if (data.timer_started)
    {
        double current_time = Util::get_current_ms();
        int seconds = 300 - (current_time - data.timer_start_time) / 1000;
        seconds = std::max(seconds, 0);
        auto seconds_string = std::to_string(seconds % 60);
        seconds_string.insert(0, 2 - seconds_string.length(), '0');

        time_label = std::to_string(seconds / 60) + ":" + seconds_string;
    }

    label_bar(height * 0.75, height * 0.25, {
                                                to_string_with_precision(data.speed, 1) + " kts",
                                                to_string_with_precision(data.heading, 1) + "°",
                                                time_label,
                                                "0:00 | 0m",
                                            });

    // START LINE DISPLAY

    double pin_pos = width * 0.15;
    double boat_pos = width * 0.85;
    double line_height = height * 0.225;
    double buoy_size = 40;

    ctx->move_to(pin_pos, line_height);
    ctx->arc(
        pin_pos,
        line_height,
        buoy_size / 2,
        0,
        2 * M_PI);

    double blink = Util::get_current_ms() /
                   250.0;

    if (data.pin_pos.is_zero())
    {
        double b = (sin(blink) + 1) / 2.0;
        ctx->set_source_rgb(b, b, b);
    }
    else
    {
        ctx->set_source_rgb(1, 1, 1);
    }
    ctx->fill();

    ctx->rectangle(
        boat_pos - buoy_size / 2,
        line_height - buoy_size / 2,
        buoy_size,
        buoy_size);

    if (data.boat_pos.is_zero())
    {
        double b = (sin(blink) + 1) / 2.0;
        ctx->set_source_rgb(b, b, b);
    }
    else
    {
        ctx->set_source_rgb(1, 1, 1);
    }
    ctx->fill();

    ctx->set_source_rgb(1, 1, 1);

    ctx->move_to(pin_pos, line_height);
    ctx->line_to(boat_pos, line_height);
    std::vector<double> dashes = {14.0, 6.0};

    ctx->set_dash(dashes, 1);
    ctx->stroke();
    ctx->unset_dash();

    // WIND ARROW
    // if (data.wind_dir == 0) {
    //     double b = (sin(blink) + 1) / 2.0;
    //     ctx->set_source_rgb(b, b, b);
    // } else {
    //     ctx->set_source_rgb(1, 1, 1);
    // }

    // double arrow_size = 20;

    // ctx->move_to(width * 0.5, height * 0.12);
    // ctx->rotate(data.wind_dir); // TODO make relative to line
    // ctx->rel_line_to(0, arrow_size * 2);
    // ctx->stroke_preserve();
    // ctx->rel_line_to(-arrow_size, -arrow_size);
    // ctx->rel_line_to(arrow_size * 2, 0);
    // ctx->rel_line_to(-arrow_size, arrow_size);

    // ctx->fill();
    ctx->set_source_rgb(1, 1, 1);

    if (!data.pin_pos.is_zero() && !data.boat_pos.is_zero() && !data.current_pos.is_zero())
    {
        double line_angle = Util::lat_lon_bearing(data.pin_pos, data.boat_pos);
        double line_dist = Util::lat_lon_dist(data.pin_pos, data.boat_pos);

        double pin_angle = Util::lat_lon_bearing(data.pin_pos, data.current_pos);
        double pin_dist = Util::lat_lon_bearing(data.pin_pos, data.current_pos);

        double line_dist_screen = boat_pos - pin_pos;

        vec2d screen_pos = {cos(pin_angle + line_angle), sin(pin_angle + line_angle)};
        screen_pos *= pin_dist * line_dist_screen / line_dist;

        double person_size = 50;
        ctx->rectangle(
            screen_pos.x - person_size / 2,
            screen_pos.y - person_size / 2,
            person_size,
            person_size);

        ctx->set_source_rgb(0, 1, 0);
        ctx->fill();
    }
}

File: python/display.py
from math import pi
from turtle import width
import cairo
from framebuffer import FrameBuffer
import time

def deg_rad(deg):
    return deg / 180 * pi

img_formats = {
    1: cairo.Format.A1,
    8: cairo.Format.A8,
    16: cairo.Format.RGB16_565,
    24: cairo.Format.RGB24,
    30: cairo.Format.RGB30,
    32: cairo.Format.ARGB32
}

class Display:
    def __init__(self):

        self.framebuffer = FrameBuffer()

        img_format = img_formats[self.framebuffer.var_info.bits_per_pixel]

        self.surface = cairo.ImageSurface(
            img_format,
            self.framebuffer.var_info.xres_virtual,
            self.framebuffer.var_info.yres_virtual,
        )

        self.ctx = cairo.Context(self.surface)


        self.ctx.select_font_face(
            "serif",
            cairo.FONT_SLANT_NORMAL,
            cairo.FONT_WEIGHT_NORMAL
        )

        self.last_update = time.time_ns()

        self.last_displayed_values = {}
        self.composite_areas = {}
        
    def flip(self):
        self.framebuffer.flip(self.surface)

    def clear(self):
        self.ctx.set_source_rgba(0,0,0,1)
        self.ctx.paint()

        self.ctx.set_source_rgba(1,1,1,1)

    def get_compass_line_length(self, deg):
        if deg % 45 == 0:
            return 20

        if deg % 10 == 0:
            return 15

        if deg % 5 == 0:
            return 10

        return 0

    def draw_speed(self):
        speed = 6.7

        if speed == self.last_displayed_values.get('speed'):
            return
        
        self.last_displayed_values['speed'] = speed

        self.ctx.identity_matrix()
        self.ctx.set_font_size(30)
        self.ctx.translate(self.surface.get_width() / 2, 100)
        self.ctx.move_to(0,0)
        label = f'{speed} kts'
        extents = self.ctx.text_extents(label)
        self.ctx.rel_move_to(-extents.width/2,extents.height/2)
        self.ctx.show_text(label)

    def composite(label):
        extents_border = 5

        def _composite(func):
            def wrapper(self, val):
                if val == self.last_displayed_values.get(label):
                    return

                self.ctx.identity_matrix()
            
                if self.composite_areas.get(label) is not None:
                    rect = self.composite_areas[label]

                    self.ctx.move_to(rect[0]-extents_border, rect[1]-extents_border)
                    self.ctx.line_to(rect[0]-extents_border, rect[3]+extents_border)
                    self.ctx.line_to(rect[2]+extents_border, rect[3]+extents_border)
                    self.ctx.line_to(rect[2]+extents_border, rect[1]-extents_border)
                    self.ctx.close_path()

                    self.ctx.set_source_rgba(0,0,0,1)
                    self.ctx.fill()

                self.ctx.set_source_rgba(1,1,1,1)

                self.composite_areas[label] = func(self, val)

                self.last_displayed_values[label] = val
            return wrapper
        return _composite

    @composite('compass')
    def draw_compass(self, current_heading):
        strokes = []
        fills = []

        compass_arc = 360 / 8
        compass_rad = 150
        compass_pos = (self.surface.get_width()/2, 400)

        self.ctx.new_path()
        self.ctx.translate(compass_pos[0],compass_pos[1])
        self.ctx.arc_negative(0,0,compass_rad,deg_rad(compass_arc-90),deg_rad(-compass_arc-90))
        strokes.append(self.ctx.copy_path())
        self.ctx.new_path()

        self.ctx.move_to(0,0)

        self.ctx.set_font_size(30)
        label = f'{current_heading}'
        extents = self.ctx.text_extents(label)
        self.ctx.rel_move_to(-extents.width/2,-compass_rad+extents.height/2+35)
        self.ctx.text_path(label + '˚')
        fills.append(self.ctx.copy_path())

        self.ctx.set_font_size(20)

        offset = current_heading % 5
        self.ctx.rotate(deg_rad(-offset - 45))
        for true_deg in range(int(-compass_arc - offset), int(compass_arc - offset + 5), 5):
            deg = true_deg + current_heading

            self.ctx.move_to(0,0)
            self.ctx.rotate(deg_rad(5))
            self.ctx.rel_move_to(0,-compass_rad-30)

            line_length = self.get_compass_line_length(deg)
            self.ctx.rel_line_to(0, line_length)
            strokes.append(self.ctx.copy_path())

            if deg % 20 == 0:
                self.ctx.rel_move_to(0,-15)

                label = str(int(deg % 360))

                extents = self.ctx.text_extents(label)
                self.ctx.rel_move_to(-extents.width/2,-extents.height/2)
                self.ctx.text_path(label)
                fills.append(self.ctx.copy_path())

        self.ctx.identity_matrix()
        for fill in fills:
            self.ctx.append_path(fill)

        self.ctx.identity_matrix()
        fill_extents = self.ctx.fill_extents()
        self.ctx.fill()
        self.ctx.identity_matrix()

        for stroke in strokes:
            self.ctx.append_path(stroke)
        self.ctx.stroke()

        self.ctx.identity_matrix()
        stroke_extents = self.ctx.path_extents()
        self.ctx.stroke()

        extents = [
            min(stroke_extents[0], fill_extents[0]),
            min(stroke_extents[1], fill_extents[1]),
            max(stroke_extents[2], fill_extents[2]),
            max(stroke_extents[3], fill_extents[3])
        ]

        return extents

    @composite('fps')
    def fps(self, fps):
        self.ctx.set_font_size(20)

        label = f'{fps:.2f} fps'
        # self.ctx.translate(20,20)
        self.ctx.move_to(20,20)
        self.ctx.text_path(label)
        extents = self.ctx.fill_extents()
        self.ctx.fill()

        self.ctx.identity_matrix()
        return extents